### ISSUE TITLE
Solution to provide Angular app deployment flexibility

### ISSUE THREAD

**Issue by ActionAnalytics** [bcgov/NRPTI#25]
Routing information is being placed in at develop time in the Angular api service files.  This makes deployment involve app code changes.  Here are the hardcoded values:

https://github.com/bcgov/NRPTI/blob/6e1ed224ac393e558cf2ca7109caa65251be3eb2/angular/projects/admin-nrced/src/app/services/api.ts#L104

**Background**

This has been a challenge to address up to now in API backed Angular front-ends ([architecture diagram](https://github.com/bcgov/eagle-dev-guides/blob/master/dev_guides/architecture.md)) that are built using the s2i built approach ([build diagram](https://github.com/bcgov/eagle-dev-guides/blob/master/dev_guides/s2i_extended_builds.md)), which creates a high performance web server image but at the cost of extra build complexity.

**Solution**

Here is a simple pattern from Eagle EPIC that solves this in a few lines, meeting the https://12factor.net Factor III, "Store config in the environment" and allowing the app to be deployed based on settings alone.  It handles the more complicated s2i builds (other projects that expose environment settings in a similar pattern like MDS use a much [simpler build process](https://github.com/bcgov/mds/blob/develop/openshift/_nodejs.bc.json)).

Please see the following areas of code:

1.  [Here's where the Angular app consumes the settings from the server and replaces the hardcoded routes](https://github.com/bcgov/eagle-public/blob/53b42222280aaad9972525ecdffd4ad4ff4b73e4/src/app/services/api.ts#L41)
1.  [Make sure this line gets placed in the HTML head as it needs to run prior to app load](https://github.com/bcgov/eagle-public/blob/1c51106bb7f94495391c9a0f322f06fd533d792e/src/index.html#L61)
1.  [Create the landing point for the environment variables at server build time](https://github.com/bcgov/eagle-public/blob/53b42222280aaad9972525ecdffd4ad4ff4b73e4/openshift/templates/nginx-runtime/Dockerfile#L7)
1. [When the pod starts up, and on every subsequent redeployment, set the variables](https://github.com/bcgov/eagle-public/blob/53b42222280aaad9972525ecdffd4ad4ff4b73e4/openshift/templates/nginx-runtime/s2i/bin/run#L5).  Javascript was chosen over JSON as it's a direct plugin for the existing variables in api.ts without having to do overhead to handle the JSON.  [MDS uses JSON](https://minesdigitalservices.gov.bc.ca/env).
1. [Pass the environment variables along](https://github.com/bcgov/eagle-public/blob/53b42222280aaad9972525ecdffd4ad4ff4b73e4/openshift/templates/angular-on-nginx/angular-on-nginx-deploy.json#L75)
1. [Settings exposed in OpenShift](https://github.com/bcgov/eagle-public/blob/53b42222280aaad9972525ecdffd4ad4ff4b73e4/openshift/templates/angular-on-nginx/angular-on-nginx-deploy.json#L252)

You can check if your settings are exposed by (if you used the same pattern as above) visiting https://your-front-end-url/publicServerEnvironmentSettings.js

And remember to set your app to redeploy on environment variable change.  That way your app will get new environment settings without needing a build.

Cheers!

PS. If you have a discrete admin site, the above solution uses the same pattern with slight variations on the path (/admin in the [server build](https://github.com/bcgov/eagle-admin/blob/d18f246f93a3b024604ea004a5193634cb865a12/openshift/templates/nginx-runtime/Dockerfile#L8) and image [startup script](https://github.com/bcgov/eagle-admin/blob/d18f246f93a3b024604ea004a5193634cb865a12/openshift/templates/nginx-runtime/s2i/bin/run#L5) but [html stays the same](https://github.com/bcgov/eagle-admin/blob/d18f246f93a3b024604ea004a5193634cb865a12/src/index.html#L46)).  Note the use of uniquely named client-side variables to avoid collisions ([public](https://github.com/bcgov/eagle-public/blob/develop/src/app/services/api.ts#L41) vs [admin](https://github.com/bcgov/eagle-admin/blob/develop/src/app/services/api.ts#L62)).

**Comment by NickPhura** [comment]
Looks interesting.  This concept in general, of loading config from environment variables (or similar) seems worthwhile.

**Comment by kelpisland** [comment]
more than worthwhile - mandatory for a resilient, cloud native deployment!!

**Comment by marklise** [comment]
@ActionAnalytics I think this is an improvement over what we have been doing in the past.  Thanks for the submission.  I have a small suggestion to make one more iteration on this, perhaps we could get a custom container image to run JS in the nginx config, allowing us to go purely ENV var based without the need to have a fake file written to disk.  See: https://www.nginx.com/blog/introduction-nginscript/  for more details.  There are many other ways to do this, such as envsubst but this is one I think might be most useful down the road.

**Comment by ActionAnalytics** [comment]
@marklise That's a good idea and I'd think a working code solution to share back to the community would be appreciated by all.  The provided imperfect plug and play solution gets us to 12factor on this item and further improving it has my full support.  Can you implement your suggestion and share it with the community?

**Comment by marklise** [comment]
FYI - We found an issue within EPIC (where this is actually deployed).  Once a client app visited the nginx server and downloaded the url details from the nginx service, it would store it in local-storage by design.  Unfortunately for the EPIC client app (admin/public/etc) in one scenario, this created a persistent problem until local-storage was cleared out because of a misconfiguration with the urls when injected into the nginx container.  Future re-deployments of the nginx container would not fix this.  We had to remove/disable this check and revert to the old method of determining which service endpoint to use because the client app would never re-check and always use the one stored in local-storage.

There doesn't appear to be any flaws in this design of injecting URLs, just the client app(s) in question need different handling done before we can re-enable this.
